### PR TITLE
Enable ProtoStream encoding for External Infinispan feature

### LINKED ISSUE #30931: Enable ProtoStream encoding for External Infinispan feature
### Description

Keycloak 26 will start using ProtoStream encoding to send data over the wire. The Infinispan server can take advantage of this encoding and enable advanced features, like queries.

This change involves updating the source code that searches for entities by using the Ickle Query and uploading the Protocol Buffer schemas into the Infinispan server.

The cache created must be configured with ProtoStream encoding.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by pruivo**
The ProtoStream schema is automatically uploaded to the Infinispan server during startup.
When the schema is updated, the indexes are updated and re-created.
Use the delete statement to delete entities when a realm is removed.
    
Fixes #30931

Depends on https://github.com/keycloak/keycloak/pull/31074

### CODE REVIEW

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (ryanemerson) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:194]: Shall we remove the Jira link as the issue doesn't change behaviour required here?

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (ryanemerson) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:208]: Isn't this the default?

**Review by ryanemerson — CHANGES_REQUESTED** [review]
We can remove `ByRealmIdConditionalRemover` as it's no longer used.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java:50]: It is ready for review: 
JIRA: https://issues.redhat.com/browse/ISPN-16279
PR: https://github.com/infinispan/infinispan/pull/12620

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java:51]: We can use the Infinispan blocking manager. 
It can cause issues with virtual threads. I discovered the hard way that `CompletableFuture.get()` pins the virtual thread in the carrier thread; if all carrier threads are blocked, the system stalls. 
Said that, I don't expect realms to be removed very often and it is probably a good idea to do the change.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:194]: I plan to change the default to be non-indexed and use the same configuration in all the caches. Does it make sense? I don't expect any user to use this code to create the caches but it is useful for testing and demos :)

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java:68]: I can avoid instantiating the field for user and client sessions. See my work in progress here: https://github.com/pruivo/keycloak/commit/64a3ac9bca53d707556db8d9ae1ea7c13ed64a47#diff-21831526a380413d0330ae4bf340621549e36b52968e4b8273d41360f10beec2R32

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java]: removed

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java]: Added an executor.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/transaction/SingleCacheAndExecutor.java
Comment (ryanemerson) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/transaction/SingleCacheAndExecutor.java:27]: Nitpick, but I think calling this `RemoteCacheAndExecutor` would be more accurate.

**Review by ryanemerson — APPROVED** [review]
Overall LGTM, just one very minor comment.

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @pruivo, I added some comments. Mostly based on some memories of issues we were solving for map store implementation.

**Review by ahus1 — COMMENTED** [review]
Thank you for the PR. No blockers, just comments and questions that I'd like to have answered before merging.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:228]: I'm not expecting ProtoStream to generate a schema with errors, but it doesn't hurt to add this.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:236]: Theoretically, no manual intervention is required if the user configures the automatic startup mode [1]. Only when the index schema is updated, do we need to re-index everything and that is done by KC. 
Documentation on configuring the caches is missing, although I'm not sure if I want to make indexing a requirement. 

[1] https://infinispan.org/docs/stable/titles/query/query.html#indexing-configuration_indexing-caches (scroll down for `Index startup mode`)

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:238]: yes, we can do that.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProviderFactory.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProviderFactory.java:99]: The idea is to have a single implementation per entity, see [1]. In this case, I'm using the same class for both entities since they are the same.

The implementation will handle all the fields that can be used for removals. 

[1] https://github.com/pruivo/keycloak/blob/2cd8e366767d55904b9722402a00c9a715dfb2e8/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/ClientSessionQueryConditionalRemover.java

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:249]: If we plan to support outside K8S, yes. We need admin permission and, by default, the operator creates an user with admin permission.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:238]: re-indexing is blocking and it will block the KC startup (on the first node). I think Infinispan can handle reads/writes while reindexing, but I'm not sure if queries return the correct value. 

Re-index is only required when the indexing schema changes. KC will trigger that and no manual intervention should be necessary.

I'll let @fax4ever comment on the Infinispan future.

Comment (ahus1) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:238]: As you write "will block the KC startup on the first node, and only happens on a schema change", this resolves it for me. Thank you for adding these details.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (ahus1) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:249]: Yes, we do support non-K8S setups, or try custom deployments of Infinispan on K8S despite our blueprints. I suggest to create a follow-up issue for this to add a paragraph to the docs in the right place that we would handle before we release KC26.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (mhajas) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:236]: Here is what I see for automatic startup mode:

`If data is volatile and the index is persistent then Infinispan clears the cache when it starts.`

Does this mean we will lose all the data on each Infinispan node restart? Or the startup procedure is triggered only when all Infinispan nodes are restarted? Another question, is index data somehow distributed across the Infinispan cluster?

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:277]: @ahus1 @mhajas added some code to compare the indexed fields of the entities and re-index if it is changed. 
Added a unit test: see `IndexSchemaChangeTest`

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (fax4ever) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java]: I confirm what @pruivo said

**Comment by ahus1** [comment]
@pruivo - talking about docs and operational procedures: If you loose the contents in the external Infinispan including the Protbuf schema that Keycloak uploaded, I assume that you'll need to restart at least one Keycloak node to ensure that it re-publishes the schema and triggers the indexing so that the indexing will work again. 

If this is the case, should we create a follow-up issue to handle this by adding listener for the `ViewChanged` so check that the schema is present? When we automate it, we don't need to write docs about this :-)

**Comment by pruivo** [comment]
> @pruivo - talking about docs and operational procedures: If you loose the contents in the external Infinispan including the Protbuf schema that Keycloak uploaded, I assume that you'll need to restart at least one Keycloak node to ensure that it re-publishes the schema and triggers the indexing so that the indexing will work again.
> 
> If this is the case, should we create a follow-up issue to handle this by adding listener for the `ViewChanged` so check that the schema is present? When we automate it, we don't need to write docs about this :-)

@ahus1 The schemas are stored in the filesystem and you would have to stop all nodes and delete the `data` directory to lose them. I'm not convinced that it is worth doing anything related to this kind of event due to its low probability.

**Comment by ahus1** [comment]
> The schemas are stored in the filesystem

Thank you, @pruivo. This is good to know and it resolves it for me. No automation is necessary for this. As I can still see that this might happen, I'd like to see the error message so we can write a short paragraph in one of the existing docs as a follow-up to this PR.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (mhajas) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:242]: I see. So the index is then created during the rebalance for all the entries that are assigned to the newly spawned node?

**Review by ahus1 — DISMISSED** [review]
Thank you @pruivo for this PR and answering all our questions. While I think this is new territory at least for me to explore when it comes to Protobuf schemas in an external Infinispan, I know I will only about it by trying it out. So I'm good to merge it and see how it evolves, as KC26 is still some time in the future. Improvements can be handled in additional PRs. 

Not merging this yet to also give Michal a chance to finish his review.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProviderFactory.java
Comment (mhajas) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProviderFactory.java:99]: Ok, that looks good to me! Thanks for sharing. 

Last question, I see that for get queries, you used a different approach than to removals (something like named queries we have for JPA): https://github.com/pruivo/keycloak/blob/2cd8e366767d55904b9722402a00c9a715dfb2e8/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/query/ClientSessionQueries.java#L33-L36

Have you considered using the same approach as for removers to have more flexibility in building clauses? I am not saying it is wrong, the advantage is that it is similar to JPA but I am curious what was the reason to have different approach than for removers.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (mhajas) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:233]: Thank @pruivo

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/LoginFailureEntity.java
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/LoginFailureEntity.java:36]: This seems to be unused? Is this some relict from other approach to reindexing per cache?

**Review by mhajas — COMMENTED** [review]
Thanks @pruivo, just last comment, otherwise it looks good to me

**Comment by ahus1** [comment]
Store model tests will fail until https://github.com/keycloak/keycloak/issues/31828 has been resolved and this PR is then rebased.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:242]: Yes. Every time a node receives a key to store (request for Keycloak, state transfer, request from another site), it indexes it.

Comment (mhajas) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:242]: Thanks for the explanation!

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProviderFactory.java
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProviderFactory.java:99]: I did not. I was trying to achieve a single delete statement with a dynamic condition. Taking the client session as an example, if you remove by realm id and by user session id in the same Keycloak transaction, it will create the condition `(realmId = :p0) || (userSessionId = :p1)`

If I had understood what you suggested, I would have 2 delete statements:

`delete from <client session> where realmId = :p0` and `delete from <client session> where userSessionId = :p0`

The disadvantage of having 2 delete statements running concurrently is, that if a key matches both conditions, you would have 2 removals concurrently happen in the cluster (and sent to the other site).

Comment (mhajas) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProviderFactory.java:99]: Yes, I understand that. I did not suggest using two queries. What I am saying is that we have different approaches for fetch queries and removes. While we have dynamic WHERE clauses creation for removes, in the commit you shared, we don't have that for fetching and we have static named queries: https://github.com/pruivo/keycloak/blob/2cd8e366767d55904b9722402a00c9a715dfb2e8/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/query/ClientSessionQueries.java#L33-L36

I was wondering why there is a different approach and whether in future we could benefit from the dynamic WHERE clauses creation also for fetching.

But this is not relevant to this PR so I think we can discuss this later if needed.

**Review by mhajas — APPROVED** [review]
Thank you @pruivo! Nice work!

**Pull request merged**
